Copy to clipboard for secret and configMap data

jhadvig opened the pull request:
Follow up of https://github.com/openshift/console/pull/23
Added:
- eye and eye-slash icon to the toggle secret visibility link
- copy to clipboard button to each value

Screens:

- Hidden secret values
![download](https://user-images.githubusercontent.com/1668218/40364633-2295a38e-5dd3-11e8-9380-539883490d56.png)


- Visible secret values
![download 1](https://user-images.githubusercontent.com/1668218/40364649-2af8757e-5dd3-11e8-85e3-d81f8635b66f.png)


- ConfigMap values
![download 2](https://user-images.githubusercontent.com/1668218/40364711-4891e584-5dd3-11e8-8bef-d351a7006ed5.png)


@spadgett @benjaminapetersen PTAL

@beanh66 @serenamarie125 fyi

spadgett commented on frontend/public/components/configmap-and-secret-data.jsx:
```diff
@@ -1,17 +1,38 @@
+    <span className="sr-only">value hidden</span>
+    <span aria-hidden="true">&bull;&bull;&bull;&bull;&bull;</span>
+  </React.Fragment>;
+}
```
Don't export items we aren't using elsewhere. Suggest just calling it `Masked`. This could be

```jsx
const Masked = <React.Fragment>
  <span className="sr-only">value hidden</span>
  <span aria-hidden="true">&bull;&bull;&bull;&bull;&bull;</span>
</React.Fragement>
```

spadgett commented on frontend/public/components/configmap-and-secret-data.jsx:
```diff
@@ -1,17 +1,40 @@
+  </React.Fragment>;
+};
+
+export const KeyValueData = ({data, showSecret}) => {
```
I have to disagree with @benjaminapetersen here... I think this was cleaner without trying to share a component for the data. They're as different as they are the same.

jhadvig replied:
@benjaminapetersen not quite sure if collapsing `ConfigMapData` and `SecretData` would do any good, since they are fairly different. For `SecretData` we need to keep the state of the `showSecret` and also we need the `toggleSecret` function that is bind to `this`. For that reason it make sense to have it as class component
On the other hand `ConfigMapData` just displays data, so making it a functional component is the right approach imho.

spadgett commented on frontend/public/components/configmap-and-secret-data.jsx:
```diff
@@ -1,17 +1,40 @@
+  return <KeyValueData data={data}/>;
+};
+
+export const Icon = ({showSecret}) => {
```
Don't export

spadgett commented on frontend/public/components/utils/_copy-to-clipboard.scss:
```diff
@@ -0,0 +1,14 @@
+.copy-to-clipboard {
```
We should be following BEM naming. Console uses a `.co-` prefix.

http://getbem.com/naming/

spadgett commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -0,0 +1,14 @@
+    <div className="copy-to-clipboard">
+      <pre className="co-pre-wrap">{visibleValue}</pre>
+      <CTC text={value}>
+        <button className="btn btn-default copy-btn" type="button"><i className="fa fa-clipboard" aria-hidden="true"></i></button>
```
Needs `sr-only` text

jhadvig commented:
@spadgett yeah I was thinking about it as well, but when I was messing about different screen widths I've figured out it could come handy in mobile, in case the block is too long and user wants to scroll to the bottom for some reason. This way he won't need to scroll back.
Just a though :)

spadgett commented on frontend/public/components/configmap-and-secret-data.jsx:
```diff
@@ -1,17 +1,40 @@
+};
+
+export const Icon = ({showSecret}) => {
+  return showSecret ? <span className="show-values"><i className="fa fa-eye-slash"></i></span> : <span className="hide-values"><i className="fa fa-eye"></i></span>;
```
Missing aria-hidden

spadgett replied:
In the mocks, the icon wasn't blue.

jhadvig replied:
Checked and the icon + link are blue only when not focused

spadgett replied:
> Checked and the icon + link are blue only when not focused

Sorry, I'm not sure I understand. The icon shouldn't be blue when not focused, at least per the original mocks.

spadgett commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -0,0 +1,14 @@
+
+import { CopyToClipboard as CTC } from 'react-copy-to-clipboard';
+
+export const CopyToClipboard = ({value, visibleValue}) => {
```
Would be nice if visibleValue was optional and it defaulted to value

spadgett commented:
> The Hide Values link looks like a button - is that the hover state ?

That's the focus outline after you click it. It's there for accessibility.

serenamarie125 commented:
I think it would be nice to add a "Copy to clipboard" tooltip and then once the copy is complete, change the tooltip to "Copied".  This is a direction we may be taking to PatternFly, so It would be nice to add this as well. @spadgett thoughts?

serenamarie125 commented:
ok thanks @spadgett couldn't tell if it was in focus or just the state after it changed, makes sense!

serenamarie125 commented:
one more request - when it's a multi line component, please have the icon be vertically top aligned, rather than center

spadgett commented:
@jhadvig I personally like how we did the button in OpenShift console. It looks weird to me if the button is 100% height. I'm not convinced the experience is really better on mobile with the tall, skinny button. It seems like it would be easy to accidentally click when scrolling.

![screen shot 2018-05-22 at 10 00 17 am](https://user-images.githubusercontent.com/1167259/40367321-068cd986-5da7-11e8-84a6-a7b497bb4c5b.png)

jhadvig commented:
@spadgett ok, will do it accordingly :+1:

spadgett commented:
> I think it would be nice to add a "Copy to clipboard" tooltip and then once the copy is complete, change the tooltip to "Copied". This is a direction we may be taking to PatternFly, so It would be nice to add this as well. @spadgett thoughts?

I think it makes sense. It would be good if we can avoid the problem on iOS where you have to tap the button twice, though. (Sometimes the first tap only triggers the tooltip.)

spadgett commented:
@kans @ggreer fyi since you reviewed #23

spadgett commented:
@jhadvig I'd suggest deferring the tooltip changes to separate PR. If we're going to switch, we should switch everywhere.

jhadvig commented:
@spadgett I've updated the PR
Regarding your https://github.com/openshift/console/pull/50#discussion_r189900836 on the icon, I have to oppose, since I've crawled trhough the console and on other places we have the icon in the same color than the link. eg:
- DC details page
![1](https://user-images.githubusercontent.com/1668218/40391074-7965806e-5e17-11e8-99e7-9e7207e725d3.png)
- BC environment tab
![2](https://user-images.githubusercontent.com/1668218/40391078-7be8db88-5e17-11e8-8c50-bc9a7a319cce.png)


Updated screen:
![3](https://user-images.githubusercontent.com/1668218/40391160-ae7e74ae-5e17-11e8-9949-7d91ae0fe0ef.png)

beanh66 commented:
@spadgett Thanks for catching that. That's my mistake on the mockup, I didn't intentionally have it as grey. I agree with @jhadvig, we would want that icon to be the same color as the link.

spadgett commented on frontend/public/components/configmap-and-secret-data.jsx:
```diff
@@ -1,14 +1,36 @@
+import { Heading, CopyToClipboard } from './utils';
 
-export const ConfigMapData = ({data}) => {
+const MaskedData = () => {
```
I think my original comment got collapsed here, but this can just be

```jsx
const MaskedData = <React.Fragment>
  <span className="sr-only">Value hidden</span>
  <span aria-hidden="true">&bull;&bull;&bull;&bull;&bull;</span>
</React.Fragment>;
```

spadgett commented on frontend/public/components/configmap-and-secret-data.jsx:
```diff
@@ -1,14 +1,36 @@
+  </React.Fragment>;
+};
+
+const Icon = ({showSecret}) => {
```
I'm not sure it's really worth pulling just the icon out here because you have to check the `showSecret` value twice... I'd keep the icon with the text.

spadgett commented on frontend/public/components/utils/_copy-to-clipboard.scss:
```diff
@@ -0,0 +1,14 @@
+.co-copy-to-clipboard {
```
The CSS is still not really following BEM. I'd expect rules like

```css
.co-copy-to-clipboard { /* ... */ }
.co-copy-to-clipboard__btn { /* ... */ }
.co-copy-to-clipboard__text { /* ... */ }
```

I'd suggest reading more about it here: http://getbem.com/introduction/

spadgett commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -0,0 +1,17 @@
+import { CopyToClipboard as CTC } from 'react-copy-to-clipboard';
+
+export const CopyToClipboard = ({value, visibleValue = value}) => {
+  return <React.Fragment>
```
You can skip the `{` and `return` and just set it to the markup directly.

```jsx
export const CopyToClipboard = ({value, visibleValue = value}) => <React.Fragment>...</React.Fragment>;
```

spadgett commented on frontend/public/components/configmap-and-secret-data.jsx:
```diff
@@ -1,14 +1,36 @@
+  return showSecret ? <span className="show-values"><i className="fa fa-eye-slash" aria-hidden="true"></i></span> : <span className="hide-values"><i className="fa fa-eye" aria-hidden="true"></i></span>;
+};
+
+export const KeyValueData = ({data, showSecret}) => {
```
I'd probably just keep this inside `SecretData`. If you keep the component, remove the `export` and update the name so it's clear it's only for secrets, not config maps.

jhadvig commented:
@spadgett comments addressed. Sorry for that.

jhadvig commented:
@spadgett sorry for the nesting... makes sense to structure this way !

spadgett commented on frontend/public/components/configmap-and-secret-data.jsx:
```diff
@@ -27,17 +32,21 @@ export class SecretData extends React.PureComponent {
+          {
+            showSecret
+              ? <React.Fragment><span><i className="fa fa-eye-slash" aria-hidden="true"></i></span> Hide Values</React.Fragment>
+              : <React.Fragment><span><i className="fa fa-eye" aria-hidden="true"></i></span> Reveal Values</React.Fragment>
```
Can you remove the spans on these two lines? I don't think they're needed

spadgett commented on frontend/public/components/configmap-and-secret-data.jsx:
```diff
@@ -1,14 +1,19 @@
-import { Heading } from './utils';
+import { Heading, CopyToClipboard } from './utils';
+
+const MaskedData = () => <React.Fragment>
```
```jsx
const MaskedData = <React.Fragment>
```

spadgett replied:
Sorry you're right

spadgett commented:
Thanks @jhadvig. Styles look good. Just a couple minor comments
